adapters/grid/grid.go

```diff
@@ -69,16 +97,222 @@ func setImpExtData(imp openrtb2.Imp) openrtb2.Imp {
+		keywords["ortb2"] = ortb2Array
+	}
+}
+
+func mergeWithReqExtKeywords(extKeywords map[string]interface{}, request *openrtb2.BidRequest) error {
+	var reqExt ReqExt
+	var reqExtKeywords map[string]interface{}
+
+	if err := json.Unmarshal(request.Ext, &reqExt); err != nil {
+		return err
+	}
+	if reqExt.Keywords != nil {
+		json.Unmarshal(reqExt.Keywords, &reqExtKeywords)
+
+		for key, keyword := range reqExtKeywords {
+			if extKeywords[key] == nil {
+				extKeywords[key] = keyword
+			} else {
+				if key == "site" || key == "user" {
+					target := extKeywords[key].(map[string]interface{})
+					from := keyword.(map[string]interface{})
+
+					for name, value := range from {
+						if target[name] == nil {
+							target[name] = value
+						} else {
+							valueArr := value.([]interface{})
+							targetArr := target[name].([]interface{})
+							target[name] = append(targetArr, valueArr...)
+						}
+					}
+				} else {
+					extKeywords[key] = keyword
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func fillSegments(publisherItem map[string]interface{}) []KeywordSegment {
+	segments := make([]KeywordSegment, 0)
+
+	for key, value := range publisherItem {
+		if key != "name" {
+			if keywords, ok := value.([]interface{}); ok {
+				for _, keyword := range keywords {
+					switch keyword.(type) {
+					case map[string]interface{}:
+						keywordSegment := keyword.(map[string]interface{})
+						if key == "segments" && keywordSegment["name"] != nil && keywordSegment["value"] != nil {
+							segment := KeywordSegment{
+								Name:  keywordSegment["name"].(string),
+								Value: keywordSegment["value"].(string),
+							}
+							segments = append(segments, segment)
+						}
+					case string:
+						segment := KeywordSegment{
+							Name:  key,
+							Value: keyword.(string),
+						}
+						segments = append(segments, segment)
+					}
+				}
+			}
+		}
+	}
+
+	return segments
+}
+
+func reformatExtKeywords(extKeywords map[string]interface{}) {
+	for name, pubData := range extKeywords {
+		if pubArr, ok := pubData.([]interface{}); ok {
+			formatedPubArr := make([]KeywordsPublisherItem, 0)
+			for _, item := range pubArr {
+				if publisherItem, itemOk := item.(map[string]interface{}); itemOk {
+					segments := fillSegments(publisherItem)
+
+					if len(segments) > 0 {
+						formatedPublisher := KeywordsPublisherItem{
+							Name:     publisherItem["name"].(string),
+							Segments: segments,
+						}
+						formatedPubArr = append(formatedPubArr, formatedPublisher)
+					}
+				}
+			}
+			if len(formatedPubArr) > 0 {
+				extKeywords[name] = formatedPubArr
+			} else {
+				delete(extKeywords, name)
+			}
+		} else {
+			delete(extKeywords, name)
+		}
+	}
+}
+
+func updateExtKeywords(keywords json.RawMessage, request *openrtb2.BidRequest) json.RawMessage {
+	var extKeywords map[string]interface{}
+	var extKWSite map[string]interface{}
+	var extKWUser map[string]interface{}
+
+	json.Unmarshal(keywords, &extKeywords)
```

Comment:
We realize an error scenario here is unlikely but we believe it's better to try to check for it and handle it